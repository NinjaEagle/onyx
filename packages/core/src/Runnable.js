/**
 * `Runnable` class that abstracts basic functionality shared
 * between the `Test` and `Suite` classes.
 * 
 * @class
 * @param {String} title
 * @param {Function} fn
 */
export default class Runnable {
  title = '';
  fn = null;
  pending = null;
  sync = null;
  async = !sync;

  constructor(title, fn) {
    this.title = title;
    this.fn = fn;
    this.async = fn && fn.length;
    this.sync = !this.async;
    this.pending = false;
  }


  /**
   * Check if this runnable or its parent suite is marked as pending.
   *
   * @private
   */
  _isPending() {
    return this.pending || this.isPending();
  }
}
