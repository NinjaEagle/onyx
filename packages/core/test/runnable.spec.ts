import Runnable from '../src/runnable';

describe('Runnable', () => {
  it('should run', () => {
    const fn = jest.fn();
    const runnable = new Runnable('test', fn);

    expect(fn).toHaveBeenCalledTimes(0);
    expect(runnable.run()).toMatchSnapshot();
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('should return isDone', () => {
    const runnable = new Runnable('test', () => null);

    expect(runnable.isDone()).toBeFalsy();
    runnable.run();
    expect(runnable.isDone()).toBeTruthy();
  });

  it('should skip', () => {
    const fn = jest.fn();
    const runnable = new Runnable('test', fn, true);

    expect(runnable.run()).toMatchSnapshot();
    expect(fn).toHaveBeenCalledTimes(0);
  });

  it('should error', () => {
    const fn = () => {
      throw new Error('Fatal error');
    };
    const runnable = new Runnable('test', fn);

    expect(runnable.run()).toMatchSnapshot();
  });

  class ExpectError extends Error {
    constructor(message: string) {
      super(message);
      this.name = 'ExpectError';
    }
  }

  it('should fail', () => {
    const fn = () => {
      throw new ExpectError('Expect failed');
    };
    const runnable = new Runnable('test', fn);

    expect(runnable.run()).toMatchSnapshot();
  });

  it('should run asynchronously', async () => {
    const fn = jest.fn();
    const runnable = new Runnable('test', fn);

    expect(fn).toHaveBeenCalledTimes(0);
    expect(await runnable.asyncRun()).toMatchSnapshot();
    expect(fn).toHaveBeenCalledTimes(1);
  });
});
