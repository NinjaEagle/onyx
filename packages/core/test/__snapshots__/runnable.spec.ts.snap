// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runnable should error 1`] = `
Result {
  "internalMessages": Array [
    "Error: Fatal error",
  ],
  "internalStatus": 2,
}
`;

exports[`Runnable should fail 1`] = `
Result {
  "internalMessages": Array [
    "ExpectError: Expect failed",
  ],
  "internalStatus": 1,
}
`;

exports[`Runnable should run 1`] = `
Result {
  "internalMessages": Array [],
  "internalStatus": 0,
}
`;

exports[`Runnable should run asynchronously 1`] = `
Result {
  "internalMessages": Array [],
  "internalStatus": 0,
}
`;

exports[`Runnable should skip 1`] = `
Result {
  "internalMessages": Array [],
  "internalStatus": 3,
}
`;
